File: 210. Course Schedule II/210. Course Schedule II.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
  const adjList = new Map();
  for (const [course, prereq] of prerequisites) {
    if (!adjList.has(course)) {
      adjList.set(course, []);
    }
    adjList.get(course).push(prereq);
  }

  const ans = [];

  const visiting = new Set();
  const visited = new Set();

  const dfs = (currentCourse) => {
    if (visiting.has(currentCourse)) {
      return false;
    }

    if (visited.has(currentCourse)) {
      return true;
    }

    visiting.add(currentCourse);

    if (adjList.has(currentCourse)) {
      for (const course of adjList.get(currentCourse)) {
        if (!dfs(course)) {
          return false;
        }
      }
    }

    visiting.delete(currentCourse);
    visited.add(currentCourse);
    ans.push(currentCourse);
    return true;
  };

  for (let course = 0; course < numCourses; course++) {
    if (!dfs(course)) {
      return [];
    }
  }

  return ans;
};

File: 210. Course Schedule II/210. Course Schedule II.py
from collections import defaultdict
from typing import List


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_list = defaultdict(list)
        for course, prereq in prerequisites:
            adj_list[course].append(prereq)

        ans = []
        visiting, visited = set(), set()

        def dfs(currentCourse):
            if currentCourse in visiting:
                return False

            if currentCourse in visited:
                return True

            visiting.add(currentCourse)

            for course in adj_list[currentCourse]:
                if not dfs(course):
                    return False

            visiting.remove(currentCourse)
            visited.add(currentCourse)
            ans.append(currentCourse)
            return True

        for course in range(numCourses):
            if not dfs(course):
                return []

        return ans
